File: bench/testdata/samples/A03/PSB-A03-XSS-001/javascript/secure.js
// PSB-A03-XSS-001: Comment display
// CWE: CWE-79
// Expected: (none - secure)
const express = require("express");
const db = require("./db");
const app = express();

function escapeHtml(str) {
  return str
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

app.get("/posts/:id/comments", async (req, res) => {
  const postId = req.params.id;
  const comments = await db.query(
    "SELECT author, text, created_at FROM comments WHERE post_id = $1 ORDER BY created_at",
    [postId]
  );

  let html = "<html><body><h1>Comments</h1>";
  for (const c of comments.rows) {
    const safeAuthor = escapeHtml(c.author);
    const safeText = escapeHtml(c.text).replace(/\n/g, "<br>");
    html += `<div class="comment">`;
    html += `<strong>${safeAuthor}</strong>`;
    html += `<span>${c.created_at}</span>`;
    html += `<p>${safeText}</p>`;
    html += `</div>`;
  }
  html += "</body></html>";
  res.send(html);
});

module.exports = app;

File: bench/testdata/samples/A03/PSB-A03-XSS-001/python/secure.py
# PSB-A03-XSS-001: Comment display
# CWE: CWE-79
# Expected: (none - secure)
from flask import Flask, g
from markupsafe import escape
import sqlite3

app = Flask(__name__)

def get_db():
    if "db" not in g:
        g.db = sqlite3.connect("app.db")
        g.db.row_factory = sqlite3.Row
    return g.db

@app.route("/posts/<int:post_id>/comments")
def show_comments(post_id):
    db = get_db()
    comments = db.execute(
        "SELECT author, text, created_at FROM comments WHERE post_id = ? ORDER BY created_at",
        (post_id,),
    ).fetchall()

    html = "<html><body><h1>Comments</h1>"
    for c in comments:
        safe_author = escape(c["author"])
        safe_text = escape(c["text"]).replace("\n", "<br>")
        html += f'<div class="comment">'
        html += f"<strong>{safe_author}</strong>"
        html += f'<span>{c["created_at"]}</span>'
        html += f"<p>{safe_text}</p>"
        html += "</div>"
    html += "</body></html>"
    return html

if __name__ == "__main__":
    app.run()
